blib/man3/vcfParse.3
.\" Automatically generated by Pod::Man 4.11 (Pod::Simple 3.35)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\" ========================================================================
.\"
.IX Title "vcfParse 3"
.TH vcfParse 3 "2018-12-10" "perl v5.26.2" "User Contributed Perl Documentation"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
vcfParse \- Easy parsing of VCF files
.SH "VERSION"
.IX Header "VERSION"
Version 0.04
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 1
\& use vcfParse;
\&
\& my $vcf_file = \*(Aqin_file.vcf\*(Aq;
\& my ( $data, $info_fields, $filtered_vars, $heads ) = vcfParse::parse($vcf_file);
\&
\& my (@headers) = @{$heads};
\&
\& for ( sort { @{ $data\->{$a}}[0] cmp @{ $data\->{$b}}[0] or
\&      @{ $data\->{$a}}[1] <=> @{ $data\->{$b}}[1]
\&    }  keys %{ $data } ){
\&    my ( $chrom, $pos, $id, $ref, $alt, $quality_score, $filt, $info_block, $format_block, $tumour_info_block, $normal_info_block, $filtered_vars, $samples ) = @{ $data\->{$_} };
\&
\&    my (@format)                  = @{ $info_fields\->{$_}[0] }; # The format field from VCF
\&          my (%format_long)  = @{ $info_fields\->{$_}[1] }; # The description of the FORMAT field (corresponding to ##FORMAT=<ID=[field] in header)
\&          my (%info_long)    = @{ $info_fields\->{$_}[2] }; # The description of the INFO field (corresponding to ##INFO=<ID=[field] in header)
\&          my (@tumour_parts) = @{ $info_fields\->{$_}[3] }; # The tumour values (for each FORMAT feild)
\&          my (@normal_parts) = @{ $info_fields\->{$_}[4] }; # The normal values (for each FORMAT feild)
\&          my (%information)  = @{ $info_fields\->{$_}[5] }; # The format field from VCF
\&          my (%sample_info)  = @{ $info_fields\->{$_}[6] }; # Per\-sample info lookup (e.g. $sample_info{$_}{\*(Aq[sample_name]\*(Aq}{\*(Aq[FORMAT_field]\*(Aq})
\&
\&    if ( $sample_info{$_}{\*(AqTUMOR\*(Aq}{\*(AqAF\*(Aq} ){
\&        print "Allele frequency = "$sample_info{$_}{\*(AqTUMOR\*(Aq}{\*(AqAF\*(Aq}\en";
\&    }
\&    # Do more things
\& }
.Ve
.SH "SUBROUTINES/METHODS"
.IX Header "SUBROUTINES/METHODS"
.SS "\fBparse()\fP"
.IX Subsection "parse()"
.SH "AUTHOR"
.IX Header "AUTHOR"
Nick Riddiford, \f(CW\*(C`nick.riddiford@curie.fr\*(C'\fR
.SH "BUGS"
.IX Header "BUGS"
Please report any bugs or feature requests to \f(CW\*(C`nick.riddiford@curie.fr\*(C'\fR,
or via the GitHub issues page <https://github.com/nriddiford/vcfParse/issues>
.SH "SUPPORT"
.IX Header "SUPPORT"
.IP "Github: <https://github.com/nriddiford/vcfParse/blob/master/README.md>" 1
.IX Item "Github: <https://github.com/nriddiford/vcfParse/blob/master/README.md>"
.PD 0
.IP "perdoc: 'perldoc vcfParse'" 1
.IX Item "perdoc: 'perldoc vcfParse'"
.PD
.SH "LICENSE AND COPYRIGHT"
.IX Header "LICENSE AND COPYRIGHT"
\&\s-1MIT\s0 License
.PP
Copyright (c) 2017 Nick Riddiford
.PP
Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the \*(L"Software\*(R"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:
.PP
The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.
.PP
\&\s-1THE SOFTWARE IS PROVIDED \*(L"AS IS\*(R", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.\s0


Changes
Revision history for vcfParse

0.01    11/08
        First version, released on an unsuspecting world.
0.02    12/09
        Remove filter section and rename $filtered_snvs to $variants
0.03    15/09
        Add @headers to $info_fields->{$_}[0]
	Bulk up README.md and test.pl to explain hash contents
0.04    15/09
        Return @headers instead (to print outside of loop)

vcfParse-0.01/lib/vcfParse.pm
package vcfParse;
use strict;
use warnings;

require Exporter;

our @ISA = qw(Exporter);

our %EXPORT_TAGS = (
        'all' => [ qw(
                        parse
                ) ]
);

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw( parse );

=head1 NAME

vcfParse - Easy parsing of VCF files

=head1 VERSION

Version 0.04

=cut

our $VERSION = '0.04';

=head1 SYNOPSIS

 use vcfParse;

 my $vcf_file = 'in_file.vcf';
 my ( $data, $info_fields, $filtered_vars, $heads ) = vcfParse::parse($vcf_file);

 my (@headers) = @{$heads};

 for ( sort { @{ $data->{$a}}[0] cmp @{ $data->{$b}}[0] or
      @{ $data->{$a}}[1] <=> @{ $data->{$b}}[1]
    }  keys %{ $data } ){
    my ( $chrom, $pos, $id, $ref, $alt, $quality_score, $filt, $info_block, $format_block, $tumour_info_block, $normal_info_block, $filtered_vars, $samples ) = @{ $data->{$_} };

    my (@format) 		  = @{ $info_fields->{$_}[0] }; # The format field from VCF
 	  my (%format_long)  = @{ $info_fields->{$_}[1] }; # The description of the FORMAT field (corresponding to ##FORMAT=<ID=[field] in header)
 	  my (%info_long)    = @{ $info_fields->{$_}[2] }; # The description of the INFO field (corresponding to ##INFO=<ID=[field] in header)
 	  my (@tumour_parts) = @{ $info_fields->{$_}[3] }; # The tumour values (for each FORMAT feild)
 	  my (@normal_parts) = @{ $info_fields->{$_}[4] }; # The normal values (for each FORMAT feild)
 	  my (%information)  = @{ $info_fields->{$_}[5] }; # The format field from VCF
 	  my (%sample_info)  = @{ $info_fields->{$_}[6] }; # Per-sample info lookup (e.g. $sample_info{$_}{'[sample_name]'}{'[FORMAT_field]'})

    if ( $sample_info{$_}{'TUMOR'}{'AF'} ){
        print "Allele frequency = "$sample_info{$_}{'TUMOR'}{'AF'}\n";
    }
    # Do more things
 }

=cut

=head1 SUBROUTINES/METHODS

=head2 parse()

=cut

sub parse {
  my ($file) = shift;

  open my $in, '<', $file or die $!;
  my @headers;

  my (%snvs, %info, %variants);
  my ($tumour_name, $control_name);
  my %format_long;
  my %info_long;
  my $filter_count;
  my @samples;
  my $replacement_id = 1;

  while(<$in>){
    chomp;

    if (/^#{2}/){
       push @headers, $_;
       #$variants{$.} = $_;

      if (/##FORMAT/){
        my ($format_long) = $_ =~ /\"(.*?)\"/;
        my ($available_format_info) = $_ =~ /ID=(.*?),/;
        $format_long{$available_format_info} = $format_long;
      }

      if (/##INFO/) {
        my ($info_long) = $_ =~ /\"(.*?)\"/;
        my ($available_info) = $_ =~ /ID=(.*?),/;
        $info_long{$available_info} = $info_long;
      }
      next;
    }

    if (/^#{1}/){
      push @headers, $_;
      #$variants{$.} = $_;
      my @split = split;
      push @samples, $_ foreach @split[9..$#split];
      $control_name = $samples[0];
      $tumour_name = $samples[1];
      next;
    }

    my @fields = split;
    my ($chr, $pos, $id, $ref, $alt, $quality_score, $filt, $info_block, $format_block, @sample_info) = @fields;

    if ( $id eq '.' ){
      $id = $replacement_id++;
    }

    my %sample_parts;

    push @{$sample_parts{$samples[$_]}}, split(/:/, $sample_info[$_]) for 0..$#samples;

    my @normal_parts   = split(/:/, $sample_info[0]);
    my @tumour_parts;

    if($sample_info[1]){
      @tumour_parts = split(/:/, $sample_info[1]);
    }

    my @format        = split(/:/, $format_block);
    my @info_parts    = split(/;/, $info_block);

    my %sample_info;

    for my $sample (@samples){
      for my $info (0..$#format){
        $sample_info{$id}{$sample}{$format[$info]} = $sample_parts{$sample}[$info];
      }
    }

    my %information;

    foreach(@info_parts){

      my ($info_key, $info_value);

      if (/=/){
        ($info_key, $info_value) = $_ =~ /(.*)=(.*)/;
      }

      else {
        ($info_key) = $_ =~ /(.*)/;
        $info_value = "TRUE";
      }
      $information{$id}{$info_key} = $info_value;
    }

    $snvs{$id} = [ @fields[0..10] ];

    $info{$id} = [ [@format], [%format_long], [%info_long], [@tumour_parts], [@normal_parts], [%information], [%sample_info] ];

    $variants{$id} = $_;

  }
  return (\%snvs, \%info, \%variants, \@headers, \@samples);
}


=head1 AUTHOR

Nick Riddiford, C<nick.riddiford@curie.fr>

=head1 BUGS

Please report any bugs or feature requests to C<nick.riddiford@curie.fr>,
or via the GitHub issues page L<https://github.com/nriddiford/vcfParse/issues>

=head1 SUPPORT

=over 1

=item Github: L<https://github.com/nriddiford/vcfParse/blob/master/README.md>

=item perdoc: 'perldoc vcfParse'

=back

=head1 LICENSE AND COPYRIGHT

MIT License

Copyright (c) 2017 Nick Riddiford

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

=cut

1;
